table/tables/mutation_checker.go

```diff
@@ -0,0 +1,244 @@
+		}
+		indexData := make([]types.Datum, 0)
+		for i, v := range decodedIndexValues {
+			fieldType := &t.Columns[indexHelperInfo.indexInfo.Columns[i].Offset].FieldType
+			datum, err := tablecodec.DecodeColumnValue(v, fieldType, sessVars.Location())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			indexData = append(indexData, datum)
+		}
+
+		if len(m.value) == 0 {
+			err = compareIndexData(sc, t.Columns, indexData, dataRemoved, indexHelperInfo.indexInfo)
+		} else {
+			err = compareIndexData(sc, t.Columns, indexData, dataAdded, indexHelperInfo.indexInfo)
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// checkRowAdditionConsistency checks whether the values of row mutations are consistent with the expected ones
+// We only check data added since a deletion of a row doesn't care about its value (and we cannot know it)
+func checkRowAdditionConsistency(sessVars *variable.SessionVars, tableColumns []*model.ColumnInfo, dataAdded []types.Datum, mutations []mutation) error {
+	rowsAdded, _ := extractRowMutations(mutations)
+	if dataAdded == nil {
+		// should be unreachable
+		return nil
+	}
+	if len(rowsAdded) == 0 {
+		return errors.New("record mutations not found for a put of a row")
+	}
+	if len(rowsAdded) > 1 {
+		// impossible to have multiple insertions of records when modifying a single row of in a table
+		logutil.BgLogger().Error("multiple row mutations added/mutated", zap.Any("rowsAdded", rowsAdded))
+		return errors.New("multiple row mutations added/mutated")
+	}
+
+	columnMap := make(map[int64]*model.ColumnInfo)
+	for _, col := range tableColumns {
+		columnMap[col.ID] = col
+	}
+
+	if err := checkRowMutationWithData(sessVars, rowsAdded[0].value, dataAdded, columnMap); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// checkRowMutationWithData checks if the given row mutation is consistent with the expected one
+// precondition: expectedData contains all fields in order
+func checkRowMutationWithData(sessVars *variable.SessionVars, mutationValue []byte, expectedData []types.Datum, columnMap map[int64]*model.ColumnInfo) error {
+	columnFieldMap := make(map[int64]*types.FieldType)
+	for id, col := range columnMap {
+		columnFieldMap[id] = &col.FieldType
+	}
+	decodedData, err := tablecodec.DecodeRowToDatumMap(mutationValue, columnFieldMap, sessVars.Location())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// NOTE: we cannot check if the decoded values contain all columns since some columns may be skipped. It can even be empty
+	// Instead, we check that decoded index values are consistent with the input row.
+
+	for columnID, decodedDatum := range decodedData {
+		inputDatum := expectedData[columnMap[columnID].Offset]
+		cmp, err := decodedDatum.CompareDatum(sessVars.StmtCtx, &inputDatum)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if cmp != 0 {
+			logutil.BgLogger().Error("inconsistent row mutation", zap.String("decoded datum", decodedDatum.String()),
```

Comment:
For the error report we could use `reporter` from https://github.com/pingcap/tidb/pull/27388, and inconsistency events could traced by the transaction event. 
/cc @MyonKeminta @longfangsong